### ISSUE TITLE
AMReX Initialize/Finalize as Context Manager

### ISSUE THREAD

**Issue by ax3l** [AMReX-Codes/pyamrex#81]
We cannot rely on the last `del`/reference drop to an object to really release and free the object in Python:
  https://stackoverflow.com/a/17644536/2719194

```py
import amrex

amrex.initialize([])
mfab = amrex.MultiFab( ... )
del mfab  # freed in CPython, but not guaranteed by Python per se
amrex.finalize()
```

This code can lead to Python holding on to `mfab` over the lifetime of `amrex.finalize()`. When using memory in an AMReX area, their actual `free` after `finalize()` will be UB and usually segfault.

Refs.:
- https://stackoverflow.com/a/17644536/2719194
- https://doc.pypy.org/en/latest/gc_info.html

**Comment by WeiqunZhang** [comment]
The Arenas in amrex is fine in this.  That's why BArena never dies.

**Comment by ax3l** [comment]
Yes, but tests that run on pure device MultiFabs should be tearing down cleanly as well before `finalize()`.
```py
import amrex

amrex.initialize([])
mfab = amrex.MultiFab(
    boxarr,
    distmap,
    num_components,
    num_ghost,
    amrex.MFInfo().set_arena(amrex.The_Device_Arena()),
)
del mfab  # freed in CPython, but not guaranteed by Python per se
amrex.finalize()
```

**Comment by WeiqunZhang** [comment]
The device arena is also fine.
```
Arena*
The_Device_Arena ()
{   
    if        (the_device_arena) {
        return the_device_arena;
    } else {
        return The_Null_Arena();
    }
}
```

**Comment by WeiqunZhang** [comment]
`The_Null_Arena` also never dies.  The device memory has been freed at that point.  As long as the memory in MultiFab is not being used, it will be fine.

**Comment by ax3l** [comment]
Hm, I get a double free after finalize on process shutdown when I keep such an `mfab` alive #30 (update: maybe that is just an unsafe pointer/lease that we handed out, DLPack might make this safer)

**Comment by ax3l** [comment]
Also, I kinda want to be sure that I have a context where all Arena (CPU/GPU) memory is free'd again. In testing, we need to run 100s of small init/finalize tests in the same process quickly after reach other - they should not leak memory after each of of these tests.

**Comment by ax3l** [comment]
Oh I see what you mean, finalize will free all memory we allocated in arenas? @WeiqunZhang

**Comment by WeiqunZhang** [comment]
Correct.  `amrex::Finalize` will free all the device memory and then delete the device arena.  After that `The_Device_Arena()` will return a pointer to the Null Arena that is a static variable in a function, and the Null Arena does not do anything when its free function is called.  In C/C++ codes, the static Null Arena is guaranteed to outlive a variable on the stack.  Even if a MultiFab is a static variable in the main function, as long as it's declared after amrex::Initialize is called, we are still fine because the Null Arena is created in amrex::Initialize and the static objects will be destroyed in reverse orders when the program terminates.

But maybe things are different when python is involved.

**Comment by ax3l** [comment]
That sounds good. I'll try to build a reproducer for the issue I see and change to not use a fixture that is cached in pytest, when I need a multifab.

Another bandage I'll apply for the pattern above is to call `gc.collect()` programmatically when `amrex.finalize()` is called, to make the above example standard compliant without even requiring for this simple case that refs can outlive the memory arena.

**Comment by ax3l** [comment]
#214 binds `FabArray::clear`, which we can use in context managers to explicitly free the memory in MultiFabs when we do not need it anymore, independent of its destructor call time :)

**Comment by sayerhs** [comment]
@ax3l is there a reason why `amrex_init` is function scope instead of session scope?

**Comment by ax3l** [comment]
@sayerhs yes, the idea is to clean it out for every test and start fresh, so they are isolated. Example: different parameters in `ParmParse` and other lurking/static state that is reset in init/finalize cycles.

**Comment by ax3l** [comment]
I started developing the idea a bit further and developed a AMReX-session safe RAII container for vectors in https://github.com/AMReX-Codes/amrex/pull/5253

I am thinking of developing with a similar shared_ptr + weak_ptr concept a kind of `AMReXSession` object that we could hold onto as simulations and depend object lifetimes on in Python, but have to ponder the idea a bit more when I have less deadlines looming.

**Comment by ax3l** [comment]
Brainstormed a bit with two LLMs on this. This comment is mostly a mental research checkpoint for myself.

There is definitely a quite verbose way to bolt this all into pyAMReX, but if we look holistic at this: AMReX + pyAMReX + downstream simulation we can do a better design that plays well together, with easier implementations and lifetime tracking for AMReX users.

Interestingly, if we elevate the book-keeping in AMReX a bit we can protect Arena-backed AMReX containers automatically for all downstream users/developers, regardless of who creates it (Python, C++ simulation, AMReX internals). Claude is especially optimistic on how small an initial patch can be (in LOC).

## Claude Opus 4.6 Max

[claude_opus_4.6.md](https://github.com/user-attachments/files/26478841/claude_opus_4.6.md)

Pro: cleaner approach, able to iterate deeper into needs we have (e.g. ParmParse, too)
Con: a bit verbose per container, but doable

## Codex GTP 5.4 High

[codex_gpt_5.4.md](https://github.com/user-attachments/files/26478838/codex_gpt_5.4.md)

Pro: caught that we should also track zero-copy memory leases.
Con: quite verbose solution, did not try to generalize approach or tie into existing Arena safety logic

**Comment by ax3l** [comment]
Just to ensure there is no confusion with my PRs I linked here in the last days: my goal for all our apps and pyAMReX is that they can never segfault or call process abort directly, because that is intrusive and terminates the controlling process, which is a) not (Python) user friendly and b) counter to many use cases (ML surrogates, optimizers, agentic AI, use AMReX/downstream codes as modular libraries, ...) we will need in the coming decade.

Fixes like https://github.com/AMReX-Codes/amrex/pull/5270 and the general Null_Arena handling are _good_ to avoid issues with double free in complicated destruction logic, but they are a measure of last resort: by the very nature of it this is just a work-around that usually leaves half invalid objects behind (e.g. memory is gone, container size is still >0), which are easy to segfault if touched again, even if just by accident (or even just by tab completion in IPython).

Thus, the larger goal to achieve here for the safety described is to make sure AMReX containers do not get into _undefined/unsafe state_ (for vectors, see AMReX-Codes/amrex#5253). They might _clear_ (empty, clean) themselves of their device data and _throw_ if accessed in uninitialized/invalidated state, but they should never become prone to _segfaults_. Since we have clean define logic in AMReX, we should even be able to reinit every container in a follow-up AMReX session of the same process.

**Comment by AlexanderSinn** [comment]
Maybe it would be easier to keep the memory arenas (+GPU context / MPI context ?) around between AMReX invocations, e.g. with a distinction between soft-finalize and a final-finalize. Or do this with nested AMReX sessions where memory is tied to the outermost session, of which there can only be one (but multiple inner sessions). In both cases individual AMReX sessions can't control the parameters of the arenas through parm parse; however, I think for that it should be possible for the user to find decent compromise settings that work for all types of applications. 

Another thought I had is that CArena, though its busylist, is effectively reference counted by all of its allocations. amrex::finalize could keep the CArenas still alive but set a flag on them. If then a free is called on the CArena after which its busy list is empty and the flag is set, then it can safely destruct itself. The next amrex::Initialize needs to reuse the old CArenas if they still exist. I'm not sure if this works with Python object lifetimes, however.
